core/editor/src/tools/mod.rs

```diff
@@ -9,14 +9,19 @@ mod sample;
 mod select;
 mod shape;
 
-use crate::events::{Event, ModKeys, MouseState, Trace, TracePoint};
+use crate::events::{Event, ModKeys, MouseState, Response, Trace, TracePoint};
 use crate::Color;
+use crate::Document;
 use crate::EditorError;
 use document_core::Operation;
 use std::collections::HashMap;
 
 pub trait Tool {
-	fn handle_input(&mut self, event: Event) -> Vec<Operation>;
+	fn handle_input(&mut self, event: &Event, document: &Document) -> (Vec<Response>, Vec<Operation>);
+}
+
+pub trait Fsm {
+	fn transition(self, event: &Event, document: &Document, responses: &mut Vec<Response>, operations: &mut Vec<Operation>) -> Self;
```

Comment:
I don't love the idea of returning data to the caller of `.transition()` by passing in mutable `Vec`s. Feels way too C-like. It's probably better that we just return a 3-tuple from `.transition()` with the two `Vec`s (transferring ownership to the caller) plus the current return value. We can destructure them in the caller.

Reply:
This might be a minor thing, but reusing vectors would yield better performance as you don't have to do heap allocations for every call because the maximum number of responses/operations for one tool should not fluctuate too much (I'd suspect something in the range of 1-10) so reusing a vec might be a good idea.

Reply:
Does this get reused? I thought it's called once per input event.

Reply:
You could store a reusable Vec in the tool definition which we currently don't do. At the moment it does not make afaik any difference, but passing the vec by ref could in the fufure provide better performance, just thought I'd mention that while we are defining traits to avoid having to change things down the line